### ISSUE TITLE
Async & Performance - Ch1 - process.nextTick()

### ISSUE THREAD

**Issue by guyonroche** [getify/You-Dont-Know-JS#543]
Great book series BTW

A word of caution about using process.nextTick to 'chunk' your work. Consider the following node app...

```javascript
var a = 0;
var count = parseInt(process.argv[2]);

setTimeout(function() {
	var b = a; // careful to get the instant 'now' version of a
	console.log(b);
}, 0);

function work() {
	if (++a < count) {
		process.nextTick(work);
	}
}

work();
```

In this code, I've chunked the work (incrementing a up to count) by calling process.nextTick and you might expect the setTimeout callback to happen sometime before the work finishes but it doesn't. Even for gigantic values of count. The problem is process.nextTick adds the function to the event queue without giving setTimeout (or any other IO for that matter) a chance to get a look in. It's like a DOS attack inside your JS program.

A slightly better function to use is setImmediate - which lets IO et al have a go before scheduling the callback.

Guyon

**Comment by getify** [comment]
Yes, decent point. Will consider a tweak for second edition. Thanks.

**Comment by t9md** [comment]
Great book thanks!

The "Jobs" part of same section "ch1 of async".

> It's kinda like saying, "oh, here's this other thing I need to do later, but make sure it happens right away before anything else can happen."

For me, above sentence and example code seems very similar to `process.nextTick` explained [here](https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#process-nexttick).

So more clarification is helpful if it's different.

Below is modified example code achieving **schedule** by `process.nextTick`.
Added log "finish" to demonstrate it's asynchronicity.

```javascript
function schedule(callback) {
  process.nextTick(callback);
}

console.log( "A" );

setTimeout( function(){
	console.log( "B" );
}, 0 );

// theoretical "Job API"
schedule( function(){
	console.log( "C" );

	schedule( function(){
		console.log( "D" );
	} );
} );

console.log('finish');

/* output of program

A
finish
C
D
B

*/
```

**Comment by t9md** [comment]
After I put above commend, I tried to find out reason why Jobs explained on "Jobs ch1.md" is so similar to `process.nextTick` of nodeJS.
What I found is both Promise and `process.nextTick` queues callback onto **miscrotask**.

So the similarity is natural consequence.

Then for me, where this book correspond `setTimeout(..0)` to `In Node.js, a similar approach is process.nextTick(..).`, it should be appropriate if `In Node.js, a similar approach is process.setImmediate(..).`.
Since `process.nextTick` is special as like Promise for it's task queue to use.
